receiver/k8sclusterreceiver/collection/metadata.go

```diff
@@ -25,42 +25,158 @@ import (
-		properties[strings.ToLower(or.Kind)] = or.Name
-		properties[strings.ToLower(or.Kind)+"_uid"] = string(or.UID)
+		metadata[strings.ToLower(or.Kind)] = or.Name
+		metadata[strings.ToLower(or.Kind)+"_uid"] = string(or.UID)
 	}
 
 	return &KubernetesMetadata{
 		resourceIDKey: getResourceIDKey(rType),
 		resourceID:    string(om.UID),
-		properties:    properties,
+		metadata:      metadata,
 	}
 }
 
 func getResourceIDKey(rType string) string {
 	return fmt.Sprintf("k8s.%s.uid", rType)
 }
+
+// mergeKubernetesMetadataMaps merges maps of string (resource id) to
+// KubernetesMetadata into a single map.
+func mergeKubernetesMetadataMaps(maps ...map[string]*KubernetesMetadata) map[string]*KubernetesMetadata {
+	out := map[string]*KubernetesMetadata{}
+	for _, m := range maps {
+		for id, km := range m {
+			out[id] = km
+		}
+	}
+
+	return out
+}
+
+// KubernetesMetadataExporter provides an interface to implement
+// ConsumeKubernetesMetadata in Exporters that support metadata.
+type KubernetesMetadataExporter interface {
+	// ConsumeKubernetesMetadata will be invoked every time there's an
+	// update to a resource that results in one or more KubernetesMetadataUpdate.
+	ConsumeKubernetesMetadata(metadata []*KubernetesMetadataUpdate) error
+}
+
+// KubernetesMetadataUpdate provides a delta view of metadata on a resource between
+// two revisions of a resource.
+type KubernetesMetadataUpdate struct {
+	// ResourceIDKey is the label key of UID label for the resource.
+	ResourceIDKey string
+	// ResourceID is the Kubernetes UID of the resource. In case of
+	// containers, this value is the container id.
+	ResourceID string
+	// MetadataToAdd contains key-value pairs that are newly added to
+	// the resource description in the current revision.
+	MetadataToAdd map[string]string
+	// MetadataToUpdate contains key-value pairs that have been updated
+	// in the current revision compared to the previous revisions(s).
+	MetadataToUpdate map[string]string
+	// MetadataToRemove contains key-value pairs that no longer describe
+	// a resource and needs to be removed.
+	MetadataToRemove map[string]string
+}
+
+// GetKubernetesMetadataUpdate processes kubernetes metadata updates and returns
+// a map of a delta of metadata mapped to each resource.
+func GetKubernetesMetadataUpdate(old, new map[string]*KubernetesMetadata) []*KubernetesMetadataUpdate {
+	var out []*KubernetesMetadataUpdate
+
+	for id, oldMetadata := range old {
+		// if an object with the same id has a previous revision, take a delta
+		// of the metadata.
+		if newMetadata, ok := new[id]; ok {
+			if toAdd, toRemove, toUpdate, isNonEmpty :=
+				getPropertiesDelta(oldMetadata.metadata, newMetadata.metadata); isNonEmpty {
+				out = append(out, &KubernetesMetadataUpdate{
+					ResourceIDKey:    oldMetadata.resourceIDKey,
+					ResourceID:       id,
+					MetadataToAdd:    toAdd,
+					MetadataToRemove: toRemove,
+					MetadataToUpdate: toUpdate,
+				})
+			}
+		}
+	}
```

Comment:
Is it also needed to check for removed keys, i.e. keys that exist in `old` but do not exist in `new` and generate `KubernetesMetadataUpdate` for such keys?

Reply:
In that case we wouldn't be interested in a `KubernetesMetadataUpdate` because the existence of a resource in a previous revision but not in the current one would most likely mean that the resource is no longer active. In such cases, we can skip the resource and still have the last metadata for the resource.